core/storage/trie/polkadot_trie/polkadot_trie_cursor_impl.cpp

```diff
@@ -101,159 +110,149 @@ namespace kagome::storage::trie {
+            OUTCOME_TRY(node, seekNodeWithValue(child));
+            current_ = node;
+            return outcome::success();
+          }
+        }
+        current_ = nullptr;
+        return outcome::success();
       }
     }
-  END:
-    return followed_path;
+    BOOST_ASSERT_MSG(false, "Unreachable");
   }
 
-  outcome::result<boost::optional<PolkadotTrieCursorImpl::NodePtr>>
-  PolkadotTrieCursorImpl::seekChildWithValue(BranchPtr node) {
-    return seekChildWithValueAfterIdx(node, -1);
-  }
-
-  outcome::result<boost::optional<PolkadotTrieCursorImpl::NodePtr>>
-  PolkadotTrieCursorImpl::seekChildWithValueAfterIdx(BranchPtr node, int8_t idx) {
-    if (node == nullptr) {
-      return boost::none;
-    }
-    do {
-      for (uint8_t i = idx + 1; i < BranchNode::kMaxChildren; i++) {
-        OUTCOME_TRY(child, trie_.retrieveChild(node, i));
+  outcome::result<PolkadotTrieCursorImpl::NodePtr>
+  PolkadotTrieCursorImpl::seekNodeWithValue(NodePtr node) {
+    while (not node->value.has_value()) {
+      if (not node->isBranch()) {
+        return Error::INVALID_NODE_TYPE;  // can't be a leaf without a value
+      }
+      auto node_as_value = std::dynamic_pointer_cast<BranchNode>(node);
+      for (uint8_t i = 0; i < BranchNode::kMaxChildren; i++) {
+        OUTCOME_TRY(child, trie_.retrieveChild(node_as_value, i));
         if (child != nullptr) {
-          last_visited_child_.emplace_back(node, i);
+          last_visited_child_.emplace_back(node_as_value, i);
           switch (child->getTrieType()) {
             case NodeType::BranchEmptyValue:
               node = std::dynamic_pointer_cast<BranchNode>(child);
-              break;
+              goto BREAK;
             case NodeType::BranchWithValue:
             case NodeType::Leaf:
-              return child;
+              return std::move(child);
             case NodeType::Special:
               return Error::INVALID_NODE_TYPE;
           }
         }
       }
-    } while (not node->value.has_value());
-    return std::dynamic_pointer_cast<PolkadotNode>(node);
+    BREAK:;
+    }
+    return node;
   }
 
-  outcome::result<bool> PolkadotTrieCursorImpl::seekUpperBound(
+  outcome::result<void> PolkadotTrieCursorImpl::seekUpperBound(
       const common::Buffer &key) {
-    OUTCOME_TRY(is_empty, seekLowerBound(key));
-    if (not is_empty) {
-      if (this->key().has_value() and this->key() == key) {
-        OUTCOME_TRY(next());
+    OUTCOME_TRY(seekLowerBound(key));
+    if (this->key().has_value() and this->key() == key) {
+      OUTCOME_TRY(next());
+    }
+    return outcome::success();
+  }
+
+  outcome::result<int8_t> PolkadotTrieCursorImpl::getChildWithMinIdx(
+      BranchPtr node, uint8_t min_idx) const {
+    for (uint8_t i = min_idx; i < BranchNode::kMaxChildren; i++) {
+      OUTCOME_TRY(child, trie_.retrieveChild(node, i));
+      if (child) {
+        return i;
       }
     }
-    return false;
+    return -1;
```

Comment:
I think is bad idea to mix signed and unsigned types. If I were you, I would use only unsigned and `BranchNode::kMaxChildren` for error (as `end()` in containers).

Reply:
Used to highlight that it may return an error. I think it may be overseen more easily otherwise.

Reply:
As you wish. It's just my opinion.